internal/admission/admission.go

```diff
@@ -59,18 +58,38 @@ func (r *Router) InjectDecoder(d *admission.Decoder) error {
 	if d == nil {
 		return fmt.Errorf("decoder cannot be nil")
 	}
-	for _, h := range r.handlers {
-		if err := h.InjectDecoder(d); err != nil {
-			return err
+	for _, hs := range r.handlers {
+		for _, h := range hs {
+			if err := h.InjectDecoder(d); err != nil {
+				return err
+			}
 		}
 	}
 	return nil
 }
 
 func (r *Router) Handle(ctx context.Context, req admission.Request) admission.Response {
-	h, ok := r.handlers[req.RequestKind.Kind]
+
+	kind := req.RequestKind
+	if kind == nil {
+		kind = &req.Kind
+	}
+
+	handlers, ok := r.handlers[kind.Kind]
 	if !ok {
-		return admission.Errored(http.StatusBadRequest, fmt.Errorf("resource %s not implemented", req.RequestKind.Kind))
+		return admission.Allowed(fmt.Sprintf("no handlers for kind: %s", kind.Kind))
+	}
+
+	var h AdmissionHandler
+	for _, handler := range handlers {
```

Comment:
nit: the longer name should be carried in the larger scope
```suggestion
	var handler AdmissionHandler
	for _, h := range handlers {
```

Reply:
> The design consideration seems to center around the trade off between the added complexity of iterating through a slice of handlers instead of creating an entirely new handler. An existing handler can be a sub-router based on the version.

I think that version would need to set explicit routes for all of the kind version permutations which seems much worse. This allows the router to source the version support information from a handler on the fly vs having to explicitly make a router for K1V1 K1V2 K1V3 etc.

I am also assuming the most common behavior will be a generic handler that hoists to the most recent version and that handlers will be spit to deal with temporary deprecation states.